tests/testthat/test-statistical_correctness.R

```diff
@@ -0,0 +1,408 @@
+
+# scale by maximum real eigenvalue and divide by demography
+contact_matrix <- contact_matrix / max(eigen(contact_matrix)$values)
+contact_matrix <- contact_matrix / demography_vector
+
+# Check final_size works with Newton solver
+test_that("Check finalsize by groups works for Polymod, newton solver", {
+  r0 <- 2.0
+
+  n_demo_grps <- length(demography_vector)
+  n_risk_grps <- 4
+
+  # prepare p_susceptibility and susceptibility
+  psusc <- matrix(
+    data = 1, nrow = n_demo_grps, ncol = n_risk_grps
+  )
+  psusc <- psusc / rowSums(psusc)
+  susc <- matrix(
+    data = 1, nrow = n_demo_grps, ncol = n_risk_grps
+  )
+
+  # prepare control
+  control <- list(
+    iterations = 10000,
+    tolerance = 1e-6
+  )
+
+  # prepare outcome with iterative solver
+  epi_outcome_iterative <- final_size(
+    r0 = r0,
+    contact_matrix = contact_matrix,
+    demography_vector = demography_vector,
+    p_susceptibility = psusc,
+    susceptibility = susc,
+    solver = "iterative",
+    control = control
+  )
+
+  # prepare outcome with newton solver
+  epi_outcome_newton <- final_size(
+    r0 = r0,
+    contact_matrix = contact_matrix,
+    demography_vector = demography_vector,
+    p_susceptibility = psusc,
+    susceptibility = susc,
+    solver = "newton",
+    control = control
+  )
+
+  # check for equivalence
+  expect_equal(
+    epi_outcome_iterative$p_infected,
+    epi_outcome_newton$p_infected,
+    tolerance = 1e-5
+  )
+})
+
+#### Check for correct final size calculation in complex data case ####
+
+# test taken from EvL
+test_that("Newton solver is correct in complex case", {
+  # make a contact matrix
+  contact_matrix <- matrix(
+    data = c(
+      5.329620e-08, 1.321156e-08, 1.832293e-08, 7.743492e-09, 5.888440e-09,
+      2.267918e-09, 1.321156e-08, 4.662496e-08, 1.574182e-08, 1.510582e-08,
+      7.943038e-09, 3.324235e-09, 1.832293e-08, 1.574182e-08, 2.331416e-08,
+      1.586565e-08, 1.146566e-08, 5.993247e-09, 7.743492e-09, 1.510582e-08,
+      1.586565e-08, 2.038011e-08, 1.221124e-08, 9.049331e-09, 5.888440e-09,
+      7.943038e-09, 1.146566e-08, 1.221124e-08, 1.545822e-08, 8.106812e-09,
+      2.267918e-09, 3.324235e-09, 5.993247e-09, 9.049331e-09, 8.106812e-09,
+      1.572736e-08
+    ),
+    nrow = 6, ncol = 6
+  )
+
+  # make a demography vector
+  demography_vector <- c(
+    10831795, 11612456, 13511496,
+    11499398, 8167102, 4587765
+  )
+
+  # get an example r0
+  r0 <- 1.3
+
+  # a p_susceptibility matrix
+  p_susc <- matrix(1, nrow(contact_matrix), 1)
+  susc <- p_susc
+
+  # prepare control
+  control <- list(
+    iterations = 10000,
+    tolerance = 1e-6
+  )
+
+  epi_outcome <- final_size(
+    r0 = r0,
+    contact_matrix = contact_matrix,
+    demography_vector = demography_vector,
+    p_susceptibility = p_susc,
+    susceptibility = susc,
+    solver = "newton",
+    control = control
+  )
+
+  expect_s3_class(
+    epi_outcome,
+    "data.frame"
+  )
```

Comment:
Do we need to have this here again? It's probably enough to have it in `test-basic.R` or `test-check_output_df.R`

Reply:
Removed here and with the Newton solver